[pkg/ottl] Add schema_url field to contexts

kernelpanic77 opened the pull request:
This PR attempts to make schemaURL for TransformContexts. 

**Description:** <Describe what has changed.>
Added a breaking change to creation of TransformContext function for logs. Also made changes to all references of the function. 

**Link to tracking Issue:** <Issue number if applicable>
#30229

kernelpanic77 committed:
[feat]: added breaking changes to newTransformContext func of log.go, for making schemaURL accessible

TylerHelmuth commented on pkg/ottl/contexts/ottllog/log.go:
```diff
@@ -208,7 +220,10 @@ func (pep *pathExpressionParser) parsePath(path ottl.Path[TransformContext]) (ot
 		} else {
 			return accessSpanID(), nil
 		}
+	case "schemaURL":
```
I believe we'll need to add special logic in the `resource` and `instrumentaiton_scope` cases.  If `path.Name() == resource` and `path.Next().Name() == "schema_url" then call `accessResourceSchemaURL`.

Something like that.

TylerHelmuth replied:
Yes

evan-bradley replied:
Sorry for jumping in a bit late here, but I think we should adjust course just a bit. I think what we have now is on the right track, but we should implement something like the following type in the `internal` package:

```golang
type SchemaURLItem interface {
    SchemaUrl() string
    SetSchemaUrl(v string)
}
```

Then make methods like this on the `TransformContext` type that implement corresponding methods on the `ResourceContext`/`InstrumentationScopeContext` interfaces like this:

```golang
func (tCtx TransformContext) GetResourceSchemaURLItem() internal.SchemaURLItem {
	return tCtx.resourceLogs
}
```
The resource and scope path parsers could then just call each of these methods to implement the necessary getters and setters similar to how we're already doing it in ottllog.

As it is, I'm worried that re-implementing this for each package will produce a lot of unnecessary duplication.

TylerHelmuth replied:
Ya sorry I haven't gotten to spend a lot time thinking about the details of this PR. Thanks for steering it correctly.

kernelpanic77 replied:
> Sorry for jumping in a bit late here, but I think we should adjust course just a bit. I think what we have now is on the right track, but we should implement something like the following type in the `internal` package:
> 
> ```go
> type SchemaURLItem interface {
>     SchemaUrl() string
>     SetSchemaUrl(v string)
> }
> ```

@evan-bradley yeah this make sense since with different ottl packages, we would probably have to write a lot of similar code. So this sort of abstraction should work for our case. 

I am bit new to the project structure, so could you point me to the module inside `internal` in which this interface could fall. 

> Then make methods like this on the `TransformContext` type that implement corresponding methods on the `ResourceContext`/`InstrumentationScopeContext` interfaces like this:
> 
> ```go
> func (tCtx TransformContext) GetResourceSchemaURLItem() internal.SchemaURLItem {
> 	return tCtx.resourceLogs
> }
> ```
> 
> The resource and scope path parsers could then just call each of these methods to implement the necessary getters and setters similar to how we're already doing it in ottllog.
> 
> As it is, I'm worried that re-implementing this for each package will produce a lot of unnecessary duplication.

But I still feel that this abstraction would not have an effect on the NewTransformContext Function which we need to change for each module in ottl.

```go
func NewTransformContext(logRecord plog.LogRecord, instrumentationScope pcommon.InstrumentationScope, resource pcommon.Resource, scopeLogs plog.ScopeLogs, resourceLogs plog.ResourceLogs) TransformContext {
	return TransformContext{
		logRecord:            logRecord,
		instrumentationScope: instrumentationScope,
		resource:             resource,
		cache:                pcommon.NewMap(),
		scopeLogs:            scopeLogs,
		resourceLogs:         resourceLogs,
	}
}
```

For instance, for ottllog we would still have to pass in resourceLogs and scopeLogs. For the path parser as well would still have to call 2 separate functions each for access schemaURL using resourceLogs and scopeLogs.

kernelpanic77 replied:
@evan-bradley @TylerHelmuth any thoughts here ?

evan-bradley replied:
Hey @kernelpanic77, sorry, I wasn't able to look at this yesterday.

>any thoughts here on where I should add schemaURLItem interface ?

Add it to a new file in `pkg/ottl/contexts/internal` called `schema.go`.

>For instance, for ottllog we would still have to pass in resourceLogs and scopeLogs. For the path parser as well would still have to call 2 separate functions each for access schemaURL using resourceLogs and scopeLogs.

You're right that we would need to add getters for the schemaURL for the resource and scope levels of the TransformContext and would need to pass e.g. resourceLogs and scopeLogs into ottllog's `NewTransformContext` function. We would need to implement the getter in each of `internal/resource.go` and `internal/scope.go`, then they could just handle the `schema_url` subpath under the `resource` and `scope` paths respectively.

I realize that's pretty wordy, let me know if any of that isn't clear.

kernelpanic77 replied:
> You're right that we would need to add getters for the schemaURL for the resource and scope levels of the TransformContext and would need to pass e.g. resourceLogs and scopeLogs into ottllog's `NewTransformContext` function. We would need to implement the getter in each of `internal/resource.go` and `internal/scope.go`, then they could just handle the `schema_url` subpath under the `resource` and `scope` paths respectively.
@evan-bradley  I am not sure if i understood this correctly, since implementing the accessSchemaURL function for `internal/resource.go` and `internal/scope.go` would require access to resoure/scope Log objects. I am not sure how we can access that.

kernelpanic77 replied:
@evan-bradley  Have a look at my latest commit and let me know if this is what you were expecting. 

I am not sure though how the `parsePath` function inside `ottllog/log.go` would change. If could add in a commit to this branch, it would be really helpful.

kernelpanic77 replied:
@evan-bradley does the latest commit contain what you were looking for ?

evan-bradley replied:
@kernelpanic77 Sorry for the delay. I think this is in the right direction, but I'd make a few changes. I left a review to help explain what I'm thinking.

kernelpanic77 committed:
[fix]: Added separate functions to access schema_url using Scope/ResourceLog

kernelpanic77 committed:
[feat]: Updated Implementation using SchemaURLItem Interface for ottllog package in contexts

evan-bradley commented on pkg/ottl/contexts/ottllog/log.go:
```diff
@@ -21,21 +21,39 @@ import (
 		cache:                pcommon.NewMap(),
+		scopeLogs:            scopeLogs,
+		resourceLogs:         resourceLogs,
+		logSchemaURLItem:     LogScopeSchemaURLItem{scopeLogs: scopeLogs},
```
```suggestion
```
We don't need to store this, `scopeLogs` and `resourceLogs` should fulfill the `SchemaURLItem` interface.

evan-bradley reviewed (commented):
Thanks for your patience. We're headed in the right direction, I left a few comments to help illustrate how I think this should work.

kernelpanic77 commented on pkg/ottl/contexts/internal/scope.go:
```diff
@@ -14,6 +14,8 @@ import (
 type InstrumentationScopeContext interface {
 	GetInstrumentationScope() pcommon.InstrumentationScope
+	GetScopeSchemaURLItem() SchemaURLItem
+	SetScopeSchemaURLItem(s SchemaURLItem)
```
Yes, @evan-bradley that makes sense. will ensure to remove this from the interface.

kernelpanic77 commented on pkg/ottl/contexts/internal/scope.go:
```diff
@@ -118,3 +120,17 @@ func accessInstrumentationScopeDroppedAttributesCount[K InstrumentationScopeCont
+		},
+		Setter: func(ctx context.Context, tCtx K, val any) error {
+			if schemaURLItem, ok := val.(SchemaURLItem); ok {
+				tCtx.SetScopeSchemaURLItem(schemaURLItem)
```
@evan-bradley, if we are getting rid of the SetScopeSchemaURLItem function, what would be the role of the Setter function ? would it directly call the setter of the SchemaURLItem Interface.

kernelpanic77 replied:
Ignore the above comment @evan-bradley.

kernelpanic77 commented on pkg/ottl/contexts/internal/scope.go:
```diff
@@ -118,3 +120,17 @@ func accessInstrumentationScopeDroppedAttributesCount[K InstrumentationScopeCont
 	}
 }
+
+func accessInstrumentationScopeSchemaURLItem[K InstrumentationScopeContext]() ottl.StandardGetSetter[K] {
```
Okay, so should I use the case "schema_url" to directly call the accessInstrumentationScopeSchemaURLItem method on the ScopePathGetterSetter ?

kernelpanic77 commented on pkg/ottl/contexts/ottllog/log.go:
```diff
@@ -55,6 +73,22 @@ func (tCtx TransformContext) getCache() pcommon.Map {
+	return tCtx.resourceLogs
+}
+
+func (tCtx TransformContext) GetScopeSchemaURLItem() internal.SchemaURLItem {
+	return tCtx.logSchemaURLItem
+}
+
+func (tCtx TransformContext) SetScopeSchemaURLItem(schema internal.SchemaURLItem) {
+	tCtx.logSchemaURLItem = schema
+}
```
@evan-bradley I am bit unsure about this, because the GetResourceSchemaURLItem, would directly give access to the entire scopeLogs object, which does not follow the the interface of internal.SchemaURLItem.

evan-bradley replied:
I think the way you have it looks good now, thanks.

kernelpanic77 commented on pkg/ottl/contexts/ottllog/log.go:
```diff
@@ -208,7 +242,18 @@ func (pep *pathExpressionParser) parsePath(path ottl.Path[TransformContext]) (ot
 		} else {
 			return accessSpanID(), nil
 		}
+	case "schema_url":
```
Yes makes sense,  resourcepath/scopepath getter setter should take care of it.

kernelpanic77 commented on pkg/ottl/contexts/ottllog/log.go:
```diff
@@ -491,3 +536,31 @@ func accessStringSpanID() ottl.StandardGetSetter[TransformContext] {
+	}
+}
+
+func accessScopeSchemaURL() ottl.StandardGetSetter[TransformContext] {
```
Yes, I understood now. I have remove the `getScopeLogs` function and changed it to `getScopeSchemaURLItem` function.

kernelpanic77 committed:
[FIX]: Fixed SchemaURLItem interface to match the function call in ScopeLogs and ResourceLogs

kernelpanic77 committed:
[FIX]: Fixed resource_test.go to match the ResourceContext interface

kernelpanic77 commented:
@evan-bradley I think this implementation should now work for ottlog. I have addressed your comments in the latest commits. Requesting your review on this.

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 commented:
@evan-bradley Lets get this PR merged this week. I think we are very close to what we need. Please give your review for the latest commit. Thanks

kernelpanic77 commented:
@evan-bradley @TylerHelmuth can we merge this PR now ?
let me know if any other changes are needed.

evan-bradley reviewed (commented):
Thanks for your continued patience. Two additional requests:

1. Can you add a changelog? You can create an entry file with `make chlog-new`.
2. I realize it will add to the diff a fair amount, but could you also implement this for traces and metrics? I think it should be fairly straightforward, and I don't want to have inconsistent paths between the signals.

evan-bradley commented:
Sorry for the delay @kernelpanic77, I have been out for the past two weeks.

kernelpanic77 committed:
Merge branch 'open-telemetry:main' into schemaurl-ottl-context

kernelpanic77 commented on pkg/ottl/contexts/internal/scope_test.go:
```diff
@@ -395,13 +395,22 @@ func createInstrumentationScope() pcommon.InstrumentationScope {
+	return r.schemaURLItem
+}
+
+func (r *instrumentationScopeContext) SetScopeSchemaURLItem(schema SchemaURLItem) {
+	r.schemaURLItem = schema
+}
+
```
Yeah, makes sense @evan-bradley. Got rid of the function.

kernelpanic77 commented on pkg/ottl/contexts/internal/scope_test.go:
@evan-bradley  should we use scopeLogs, scopeMetrics, scopeTraces for creating a dummy schemaURLItem in  test ?

kernelpanic77 commented on pkg/ottl/contexts/internal/resource.go:
```diff
@@ -84,3 +85,17 @@ func accessResourceDroppedAttributesCount[K ResourceContext]() ottl.StandardGetS
 	}
 }
+
+func accessResourceSchemaURLItem[K ResourceContext]() ottl.StandardGetSetter[K] {
```
@evan-bradley should we use scopeLogs, scopeMetrics, scopeTraces for creating a dummy schemaURLItem in `internal/resource_test.go`?

evan-bradley replied:
I would just create a dummy SchemaURLItem struct that returns something we can mock against. Those are all in different packages which complicates the dependency graph a little bit, and they're probably heavier than what we need without much benefit.

kernelpanic77 commented:
> Sorry for the delay @kernelpanic77, I have been out for the past two weeks.

No issues @evan-bradley. I have added some of my questions regarding tests in the previous thread. Let me know what you think about it.

kernelpanic77 committed:
[feat]: completed implementation for spans and traces

evan-bradley commented on .chloggen/schemaurl-ottl-context.yaml:
Could you fill out the changelog?

evan-bradley reviewed (commented):
The changes you've made look good to me. Could you also update the paths in these two readmes to include the new paths?
- https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/contexts/ottlresource#paths
- https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/contexts/ottlscope#paths

kernelpanic77 committed:
Update schemaurl-ottl-context.yaml

Update changelog for final merge to include schema_url field in pkg/ottlcontext

kernelpanic77 committed:
Merge branch 'open-telemetry:main' into schemaurl-ottl-context

evan-bradley commented on .chloggen/schemaurl-ottl-context.yaml:
```diff
@@ -0,0 +1,27 @@
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: [user, api]
```
```suggestion
change_logs: [user]
```
This is breaking for the API and an enhancement for users. Can you create an additional changelog entry to capture the breaking API change?

evan-bradley commented on .chloggen/schemaurl-ottl-context.yaml:
```diff
@@ -0,0 +1,27 @@
+change_type: 'enhancement'
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: 'pkg/ottl/ottlcontext'
```
```suggestion
component: 'pkg/ottl'
```

evan-bradley commented on .chloggen/schemaurl-ottl-context.yaml:
```diff
@@ -0,0 +1,27 @@
+component: 'pkg/ottl/ottlcontext'
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "The following change allows access to schema_urls fields from ottlcontexts, this enhancing flexibility and functionality of ottlcontexts"
```
```suggestion
note: "Add a `schema_url` field to access the SchemaURL in resources and scopes on all signals"
```

evan-bradley commented on pkg/ottl/contexts/internal/resource_test.go:
Sorry, I missed this one previously. Can you test the `schema_url` field here as well?

kernelpanic77 committed:
[fix]: separated changelogs for api breaking changes and user enhancement changes

evan-bradley commented:
@kernelpanic77 Can you verify that you have pushed all changes in your local branch? It looks like the diff hasn't changed.

kernelpanic77 committed:
Merge branch 'schemaurl-ottl-context' of github.com:kernelpanic77/opentelemetry-collector-contrib into schemaurl-ottl-context

kernelpanic77 commented:
> @kernelpanic77 Can you verify that you have pushed all changes in your local branch? It looks like the diff hasn't changed.

Yes I think I forgot to push my local changes. Could you have a look ? I want to discuss an issue I am facing while implementing tests in `scope_test.go`, I have added comments for the same.

kernelpanic77 committed:
[feat]: add small changes to verify requirements for test_suite for scopes

evan-bradley commented on pkg/ottl/contexts/internal/resource_test.go:
Could you add a test for the `schema_url` path to this file?

evan-bradley commented on pkg/ottl/contexts/ottlresource/resource.go:
```diff
@@ -38,6 +38,11 @@ func (tCtx TransformContext) getCache() pcommon.Map {
 	return tCtx.cache
 }
 
+// [TODO]: Find appropriate way to access schemaURLItem using resource
```
My suggestion for solving this would be to move `SchemaURLItem` somewhere where it is exported, then add a new `SchemaURLItem` parameter to `NewTransformContext`. Callers can then pass in the pdata object for whatever signal they are working with.

kernelpanic77 committed:
[feat]: added breaking change to NewTransformContext function for ottlresource

kernelpanic77 committed:
[fix]: Completed Implementation TransformContext interface

kernelpanic77 committed:
[feat]: Completed the resource_test pkg.contexts.internal

kernelpanic77 committed:
[fix]: added license to the schema.go for ottl.contexts.internal

kernelpanic77 committed:
[feat]: Completed all required changes for PR

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

evan-bradley commented on pkg/ottl/contexts/ottlscope/scope.go:
```diff
@@ -45,6 +45,16 @@ func (tCtx TransformContext) getCache() pcommon.Map {
 	return tCtx.cache
 }
 
+// [TODO]:
```
Could you complete these todos as well? Should be similar to resource.

kernelpanic77 committed:
[feat]: Completed implementation for ottlscope

evan-bradley commented on pkg/ottl/contexts/internal/schema.go:
```diff
@@ -0,0 +1,8 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal
```
```suggestion
package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal"
```

evan-bradley committed:
Update pkg/ottl/contexts/internal/schema.go

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 commented:
@evan-bradley  @TylerHelmuth Does this PR require any more changes ? I think it ready to be merged. 
Let me know what you think.

evan-bradley commented on pkg/ottl/contexts/internal/schema.go:
One more thing, could you move this to a new package called `ottlcommon` under the `contexts` package? I just realized that since we're exposing this type publicly, we should make it accessible by users.

evan-bradley reviewed (approved):
Thanks for taking care of this.

@TylerHelmuth I think this should be ready to go.

TylerHelmuth commented on processor/transformprocessor/internal/common/metrics.go:
```diff
@@ -103,7 +103,7 @@ func (d dataPointStatements) ConsumeMetrics(ctx context.Context, md pmetric.Metr
 func (d dataPointStatements) handleNumberDataPoints(ctx context.Context, dps pmetric.NumberDataPointSlice, metric pmetric.Metric, metrics pmetric.MetricSlice, is pcommon.InstrumentationScope, resource pcommon.Resource) error {
 	for i := 0; i < dps.Len(); i++ {
-		tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource)
+    tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource, pmetric.NewScopeMetrics(), pmetric.NewResourceMetrics())
```
We need to pass the actual ScopeMetrics and ResourceMetrics here, otherwise accessing the schema url wont work for number datapoints.

TylerHelmuth commented on processor/transformprocessor/internal/common/metrics.go:
```diff
@@ -120,7 +120,7 @@ func (d dataPointStatements) handleNumberDataPoints(ctx context.Context, dps pme
 func (d dataPointStatements) handleHistogramDataPoints(ctx context.Context, dps pmetric.HistogramDataPointSlice, metric pmetric.Metric, metrics pmetric.MetricSlice, is pcommon.InstrumentationScope, resource pcommon.Resource) error {
 	for i := 0; i < dps.Len(); i++ {
-		tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource)
+    tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource, pmetric.NewScopeMetrics(), pmetric.NewResourceMetrics())
```
We need to pass the actual ScopeMetrics and ResourceMetrics here, otherwise accessing the schema url wont work for histogram datapoints.

TylerHelmuth commented on processor/transformprocessor/internal/common/metrics.go:
```diff
@@ -137,7 +137,7 @@ func (d dataPointStatements) handleHistogramDataPoints(ctx context.Context, dps
 func (d dataPointStatements) handleExponetialHistogramDataPoints(ctx context.Context, dps pmetric.ExponentialHistogramDataPointSlice, metric pmetric.Metric, metrics pmetric.MetricSlice, is pcommon.InstrumentationScope, resource pcommon.Resource) error {
 	for i := 0; i < dps.Len(); i++ {
-		tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource)
+		tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource, pmetric.NewScopeMetrics(), pmetric.NewResourceMetrics())
```
We need to pass the actual ScopeMetrics and ResourceMetrics here, otherwise accessing the schema url wont work for expo histogram datapoints.

TylerHelmuth commented on processor/transformprocessor/internal/common/metrics.go:
```diff
@@ -154,7 +154,7 @@ func (d dataPointStatements) handleExponetialHistogramDataPoints(ctx context.Con
 func (d dataPointStatements) handleSummaryDataPoints(ctx context.Context, dps pmetric.SummaryDataPointSlice, metric pmetric.Metric, metrics pmetric.MetricSlice, is pcommon.InstrumentationScope, resource pcommon.Resource) error {
 	for i := 0; i < dps.Len(); i++ {
-		tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource)
+    tCtx := ottldatapoint.NewTransformContext(dps.At(i), metric, metrics, is, resource, pmetric.NewScopeMetrics(), pmetric.NewResourceMetrics())
```
We need to pass the actual ScopeMetrics and ResourceMetrics here, otherwise accessing the schema url wont work for summary datapoints.

TylerHelmuth commented on processor/transformprocessor/internal/common/logs.go:
Please update each signal's `processor_test.go` with a new test that gets and sets `schema_url`.

TylerHelmuth reviewed (commented):
@kernelpanic77 left a couple comments about the transformprocessor but the implementation in the OTTL pkg looks good.

kernelpanic77 committed:
[fix]: Moved SchemaURL to ottlcommon package to expose the struct to users

kernelpanic77 committed:
[fix]: passed correct smetrics, rmetrics for ottldatapoint functions, this would allow access to correct schemaURL

kernelpanic77 committed:
[feat]: added tests to transformprocessor package to set schema_url

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

evan-bradley commented on pkg/ottl/contexts/ottlcommon/schema.go:
```diff
@@ -0,0 +1,8 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+package ottlcommon // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlcommon
```
```suggestion
package ottlcommon // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlcommon"
```

evan-bradley committed:
Update pkg/ottl/contexts/ottlcommon/schema.go

evan-bradley commented:
@kernelpanic77 Could you fix the lint issues?

TylerHelmuth commented:
running `make fmt` in the transformprocessor dir should fix the issue.

kernelpanic77 committed:
[fix]: Fixed linting issues inside transform processor

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 commented:
Ready to merge 👍

TylerHelmuth commented:
@kernelpanic77 looks like the linter is still failing. You can run `make lint` within the `pkg/ottl` dir to test if lint is passing.

kernelpanic77 commented:
> @kernelpanic77 looks like the linter is still failing. You can run `make lint` within the `pkg/ottl` dir to test if lint is passing.

@TylerHelmuth the linter is expected to fail here, since the SchemaURLItem interface uses the same signature to get and set  a schema_url string.

`Error: ottl/contexts/internal/scope_test.go:414:29: var-naming: method SchemaUrl should be SchemaURL (revive)
func (t *TestSchemaURLItem) SchemaUrl() string {
                            ^
Error: ottl/contexts/internal/resource.go:95:16: unused-parameter: parameter 'ctx' seems to be unused, consider removing or renaming it as _ (revive)
		Getter: func(ctx context.Context, tCtx K) (any, error) {
		             ^
Error: ottl/contexts/internal/scope_test.go:418:29: var-naming: method SetSchemaUrl should be SetSchemaURL (revive)
func (t *TestSchemaURLItem) SetSchemaUrl(v string) {
                            ^
Error: ottl/contexts/internal/resource_test.go:48:19: unused-parameter: parameter 'resource' seems to be unused, consider removing or renaming it as _ (revive)
			modified: func(resource pcommon.Resource) {
			               ^
Error: ottl/contexts/internal/resource_test.go:387:2: var-naming: don't use underscores in Go names; struct field schema_url should be schemaURL (revive)
	schema_url string
	^
Error: ottl/contexts/internal/resource_test.go:390:37: var-naming: method SchemaUrl should be SchemaURL (revive)
func (t *TestResourceSchemaURLItem) SchemaUrl() string {
                                    ^
Error: ottl/contexts/internal/resource_test.go:394:37: var-naming: method SetSchemaUrl should be SetSchemaURL (revive)
func (t *TestResourceSchemaURLItem) SetSchemaUrl(v string) {`

For instance `plog.Resourcelogs` also have the same signature to fetch schema_url `SchemaUrl()`

kernelpanic77 committed:
[fix]: linting errors removed from ottl/contexts

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

evan-bradley commented on pkg/ottl/contexts/ottlcommon/schema.go:
```diff
@@ -0,0 +1,8 @@
+package ottlcommon // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlcommon"
+
+type SchemaURLItem interface {
+	SchemaUrl() string
+	SetSchemaUrl(v string)
```
```suggestion
	SchemaURL() string
	SetSchemaURL(v string)
```

This is what's causing the lint issue. You'll need to update this and all calls to the interface.

TylerHelmuth replied:
@kernelpanic77 it looks like @evan-bradley's change is still required.

kernelpanic77 committed:
[fix]: Adding linting rule in golang to prevent lint on interface by gofmt

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 committed:
Removed ottlcommon package, did required changes for referencing schemaURLItem correctly

kernelpanic77 committed:
Update go.sum

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

TylerHelmuth committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 commented:
Can we merge @evan-bradley @TylerHelmuth  ?

evan-bradley commented:
@kernelpanic77 We should be able to merge this soon. Can you take a look at the failing tests?

kernelpanic77 committed:
Updated test files to fix errors in unit tests

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 committed:
Updated test files to fix errors in processort/routingprocessor

kernelpanic77 commented:
@evan-bradley I have fixed most of the integration-tests and unit-tests which seemed to be failing due to my changes. For the listing tests I think we can ignore. Can we please proceed with the merge now ?

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

kernelpanic77 committed:
Merge branch 'main' into schemaurl-ottl-context

evan-bradley commented on pkg/ottl/contexts/internal/schema.go:
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: Apache-2.0
+package internal // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/internal"
+
+//nolint:gofmt
+type SchemaURLItem interface {
+	SchemaUrl() string
+	SetSchemaUrl(v string)
+}
```
```suggestion
//revive:disable:var-naming The methods in this interface are defined by pdata types.
type SchemaURLItem interface {
	SchemaUrl() string
	SetSchemaUrl(v string)
}
//revive:enable:var-naming
```

evan-bradley committed:
Update pkg/ottl/contexts/internal/schema.go

evan-bradley commented on pkg/ottl/contexts/internal/resource_test.go:
```diff
@@ -370,14 +382,37 @@ func createResource() pcommon.Resource {
+	schemaURL string
+}
+
+func (t *TestResourceSchemaURLItem) SchemaUrl() string {
+	return t.schemaURL
+}
+
+func (t *TestResourceSchemaURLItem) SetSchemaUrl(v string) {
+	t.schemaURL = v
+}
```
```suggestion
//revive:disable:var-naming This must implement the SchemaURL interface.
func (t *TestResourceSchemaURLItem) SchemaUrl() string {
	return t.schemaURL
}

func (t *TestResourceSchemaURLItem) SetSchemaUrl(v string) {
	t.schemaURL = v
}
//revive:enable:var-naming
```

evan-bradley committed:
Fix lint

evan-bradley committed:
Run gofmt

evan-bradley commented:
@TylerHelmuth please take a look.
